esmvalcore/preprocessor/_multimodel.py

```diff
@@ -201,158 +185,88 @@ def _put_in_cube(template_cube, cube_data, statistic, t_axis):
+        cube.coord('time').units = t_unit
+        cube.coord('time').bounds = None
+        cube.coord('time').guess_bounds()
 
 
-def _assemble_full_data(cubes, statistic):
-    """Get statistical data in iris cubes for FULL."""
-    # all times, new MONTHLY data time axis
-    time_axis = [float(fl) for fl in _monthly_t(cubes)]
+def _get_time_slice(cubes, time):
+    """Fill time slice array with cubes' data if time in cube, else mask."""
+    time_slice = []
+    for cube in cubes:
+        cube_time = cube.coord('time').points
+        if time in cube_time:
+            idx = int(np.argwhere(cube_time == time))
+            subset = cube[idx].data
+        else:
+            subset = np.ma.empty(list(cube.shape[1:]))
+            subset.mask = True
+        time_slice.append(subset)
+    return time_slice
 
-    # new big time-slice array shape
-    new_shape = [len(time_axis)] + list(cubes[0].shape[1:])
 
-    # assemble an array to hold all time data
-    # for all cubes; shape is (ncubes,(plev), lat, lon)
-    new_arr = np.ma.empty([len(cubes)] + list(new_shape[1:]))
+def _assemble_data(cubes, statistic, span='overlap'):
+    """Get statistical data in iris cubes."""
+    # New time array representing the union or intersection of all cubes
+    time_spans = [cube.coord('time').points for cube in cubes]
+    if span == 'overlap':
+        new_times = reduce(np.intersect1d, time_spans)
+    elif span == 'full':
+        new_times = reduce(np.union1d, time_spans)
+    n_times = len(new_times)
 
-    # data array for stats computation
-    stats_dats = np.ma.zeros(new_shape)
+    # Target array to populate with computed statistics
+    new_shape = [n_times] + list(cubes[0].shape[1:])
+    stats_data = np.ma.zeros(new_shape)
 
-    # assemble indices list to chop new_arr on
-    indices_list = []
+    # Realize all cubes at once instead of separately for each time slice
+    _ = [cube.data for cube in cubes]
 
-    # empty data array to hold time slices
-    empty_arr = np.ma.empty(new_shape)
+    # Make time slices and compute stats
+    for i, time in enumerate(new_times):
+        time_data = _get_time_slice(cubes, time)
+        stats_data[i] = _compute_statistic(time_data, statistic)
 
-    # loop through cubes and populate empty_arr with points
-    for cube in cubes:
-        time_redone = _datetime_to_int_days(cube)
-        oidx = [time_axis.index(s) for s in time_redone]
-        indices_list.append(oidx)
-    for i in range(new_shape[0]):
-        # hold time slices only
-        new_datas_array = _full_time_slice(cubes, empty_arr, indices_list,
-                                           new_arr, i)
-        # list to hold time slices
-        time_data = []
-        for j in range(len(cubes)):
-            time_data.append(new_datas_array[j])
-        stats_dats[i] = _compute_statistic(time_data, statistic)
-    stats_cube = _put_in_cube(cubes[0], stats_dats, statistic, time_axis)
+    template = cubes[0][:n_times]
```

Comment:
why slice cube by time?

Reply:
The way the current function works is that a new cube is constructed by copying one of the input cubes and filling it with new data. This slice ensures that the length of the time dimension of the output cube corresponds with the length of the time overlap of all cubes.

Reply:
That's not quite what happens. Information from the template cube is used to construct a new cube, but its time coordinate is not used.

Reply:
No, but the length of the time coordinate is what's important here. The actual points will be overwritten by the new time coordinate, which is the same length as `n_times`.

Reply:
BTW, this is not new in this PR, so maybe we could improve it in a separate PR.

Reply:
The length of the time coordinate does not matter, just try with `template = cubes[0][:1]`.

Reply:
You're right, my bad! Fixed now.